blob/service.go

```diff
@@ -0,0 +1,274 @@
+		headerGetter: headGetter,
+	}
+}
+
+// Submit sends PFB transaction and reports the height in which it was included.
+// Allows sending multiple Blobs atomically synchronously.
+// Uses default wallet registered on the Node.
+func (s *Service) Submit(ctx context.Context, blobs ...*Blob) (uint64, error) {
+	log.Debugw("submitting blobs", "amount", len(blobs))
+
+	var (
+		gasLimit = estimateGas(blobs...)
+		fee      = int64(defaultMinGasPrice * float64(gasLimit))
+		b        = make([]*apptypes.Blob, len(blobs))
+	)
+
+	for i, blob := range blobs {
+		b[i] = &blob.Blob
+	}
+
+	resp, err := s.accessor.SubmitPayForBlob(ctx, types.NewInt(fee), gasLimit, b...)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(resp.Height), nil
+}
+
+// Get retrieves all the blobs for given namespaces at the given height by commitment.
+func (s *Service) Get(ctx context.Context, height uint64, nID namespace.ID, commitment Commitment) (*Blob, error) {
+	blob, _, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return blob, nil
+}
+
+// GetProof retrieves all blobs in the given namespaces at the given height by commitment
+// and returns their Proof.
+func (s *Service) GetProof(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Proof, error) {
+	_, proof, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+// GetAll returns all blobs under the given namespaces at the given height.
+// GetAll can return blobs and an error in case if some requests failed.
+func (s *Service) GetAll(ctx context.Context, height uint64, nIDs ...namespace.ID) ([]*Blob, error) {
+	header, err := s.headerGetter(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		blobs     = make([]*Blob, 0)
+		blobCh    = make(chan []*Blob)
+		resultErr = make([]error, 0, len(nIDs))
+	)
+
+	for i, nID := range nIDs {
+		i := i
+		nID := nID
+		go func() {
+			log.Infow("requesting blobs", "height", height, "nID", nID.String())
+			blobs, err := s.getBlobs(ctx, nID, header.DAH)
+			if err != nil {
+				resultErr[i] = fmt.Errorf("could not get the blob for the nID:%s. %s", nID.String(), err)
+				blobCh <- nil
+				return
+			}
+			blobCh <- blobs
+		}()
+	}
+
+	for range nIDs {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case b := <-blobCh:
+			if b == nil {
+				continue
+			}
+			blobs = append(blobs, b...)
+		}
+	}
+
+	if len(blobs) == 0 {
+		resultErr = append(resultErr, ErrBlobNotFound)
+	}
+	return blobs, errors.Join(resultErr...)
+}
+
+// Included verifies that the blob was included in a specific height.
+// To ensure that blob was included in a specific height, we need:
+// 1. verify the provided commitment by recomputing it;
+// 2. verify the provided Proof against subtree roots that were used in 1.;
+func (s *Service) Included(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	_ *Proof,
```

Comment:
I'm not sure that it is necessary in this case, just because the purpose of `Included` is to ensure that the requested blob was included in the requested height. In the scope of it, we should not compare proofs IMO

Reply:
I agree that to prove the inclusion of the commitment, it's okay with the current implementation to check the Proof. However, the Proof could be malformed, and the user should know this.

This will also simply be testing in Swamp. You can make a full roundtrip test that gets proofs and then verifies it. It's also not hard to do verification from how I see it

Reply:
to compare we can use reflection, but it's not the best solution for the production code. Also, please note in v1 this extra check will not make sense, so it will be removed. To summarise, I still have concerns regarding comparing proofs

Reply:
Why do you need reflection for that?

Reply:
to compare all fields (as nmt.Proof is a struct)